lib/block-supports/spacing.php

```diff
@@ -107,18 +107,23 @@ function gutenberg_render_spacing_gap_support( $block_content, $block ) {
 	}
 
 	$gap_value = $block['attrs']['style']['spacing']['blockGap'];
-
-	// Skip if gap value contains unsupported characters.
-	// Regex for CSS value borrowed from `safecss_filter_attr`, and used here
-	// because we only want to match against the value, not the CSS attribute.
-	if ( preg_match( '%[\\\(&=}]|/\*%', $gap_value ) ) {
-		return $block_content;
+	$styles = [];
+
+	if ( is_array( $gap_value ) ) {
+		foreach( $gap_value as $key => $value ) {
+			// Skip if gap value contains unsupported characters.
+			// Regex for CSS value borrowed from `safecss_filter_attr`, and used here
+			// because we only want to match against the value, not the CSS attribute.
+			if ( ! preg_match( '%[\\\(&=}]|/\*%', $value ) ) {
+				$styles[] = sprintf( '%s-gap: %s', $key, $value );
+			}
+		}
+	} else {
+		if ( ! preg_match( '%[\\\(&=}]|/\*%', $gap_value ) ) {
+			$styles[] = sprintf( 'gap: %s', $gap_value );
```

Comment:
I know this is a WIP (and you're probably already well across much of this), but just wanted to raise that a couple of the reasons we went with rendering the CSS variable inline to the block instead of using the `gap` values directly were:

* Some blocks, e.g. Columns use margins between the individual elements. Using a CSS variable for the block support means that we don't force the blocks that opt-in to use any one particular approach for handling gap.
* For blocks that might have some more DOM nesting, we don't need to skip serialization if we use a CSS variable as the variable will flow down to the children.

However, a downside of setting the variable is that other styles that also use the variable will pick it up, which can cause an issue with margins. In the PR to opt-in the [Columns block](https://github.com/WordPress/gutenberg/pull/34630) the workaround was to also opt-in to margin support so that it can be tweaked separately from the block spacing. This isn't really _ideal_ though!

Reply:
Hm, there's definitely a lot of pros/cons to balance here. The issue with syncing the single `gap` variable with the individual `row-gap` and `column-gap` ones is a big headache that I'm still trying to reason through, and the inheritance issues are another big one.

Are you suggesting going back to using the `gap` values directly? Or specifically when supporting row/column gap separately?

Reply:
> Hm, there's definitely a lot of pros/cons to balance here.

Totally!

> Are you suggesting going back to using the gap values directly? Or specifically when supporting row/column gap separately?

Apologies, no, I'm not really suggesting one way or the other. I was mostly curious to explore the trade-offs and the potential gotchas we're going to run into. I still think there's a lot of benefit to using the CSS variables (and on balance, hopefully it'll give us a bit more flexibility), but either way we're going to have a few edge cases to wrangle. So mostly I'm curious to hear what you think from the exploration!